From 237a0e5e7b1489189950746840f315e5ba612b80 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Wed, 8 Oct 2025 17:23:16 +0200
Subject: [PATCH] usb: dwc2: add support for runtime PM

In order to properly manage pm-domains and low power modes on stm32,
allow the pm runtime in the dwc2 driver. Basically, every time the bus
gets suspended (either as host, or as device), the suspend interrupt will
put the controller into one of the low power modes. On stm32, as
DWC2_POWER_DOWN_PARAM_NONE is set, the driver puts the controller in
clock gating. Similarly, upon resume (by the bus, or the local
software), the controller is taken out of clock gating.

So upon entry/exit of the clock gating mode, add pm_runtime_put/get calls
so the runtime status represents this low power mode. It manages the
power domain accordingly.
A specific case is added in drd.c, when a cable gets detached, the none
role gets selected. In such case, the controller can also be put into
low power, and runtime suspended state.
When calling dwc2_hsotg_core_disconnect(), the controller has already
been resumed. So remove un-necessary exit of low power there.
At initialization time, in the gadget init, add a runtime_get call.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I21851f9a71a09d7aa08b1b9ebb89255bb3b570ca
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512989
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/usb/dwc2/core.h     |   1 +
 drivers/usb/dwc2/drd.c      |  28 ++++++++++
 drivers/usb/dwc2/gadget.c   | 106 +++++++++++++++++++++++++++---------
 drivers/usb/dwc2/hcd.c      |  23 ++++++++
 drivers/usb/dwc2/platform.c |  12 ++++
 5 files changed, 144 insertions(+), 26 deletions(-)

--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -1076,6 +1076,7 @@ struct dwc2_hsotg {
 	unsigned int hibernated:1;
 	unsigned int in_ppd:1;
 	bool bus_suspended;
+	bool rpm_suspended;
 	enum usb_device_state suspended_from;
 	unsigned int reset_phy_on_wake:1;
 	unsigned int need_phy_for_wake:1;
--- a/drivers/usb/dwc2/drd.c
+++ b/drivers/usb/dwc2/drd.c
@@ -247,6 +247,34 @@ skip:
 		role == USB_ROLE_NONE ? "No" :
 		role == USB_ROLE_HOST ? "A" : "B");
 
+	spin_lock_irqsave(&hsotg->lock, flags);
+	if (role == USB_ROLE_NONE &&
+	    (hsotg->lx_state == DWC2_L0 || hsotg->lx_state == DWC2_L3)) {
+		/*
+		 * With none role, the controller can be put in low power mode,
+		 * in case of a new event, e.g. usb-role-switch, role_sw_set
+		 * will be called, to resume.
+		 */
+		switch (hsotg->params.power_down) {
+			int ret;
+		case DWC2_POWER_DOWN_PARAM_PARTIAL:
+			ret = dwc2_enter_partial_power_down(hsotg);
+			if (ret)
+				dev_err(hsotg->dev, "enter partial_power_down failed %d\n", ret);
+			spin_unlock_irqrestore(&hsotg->lock, flags);
+			return ret;
+
+		case DWC2_POWER_DOWN_PARAM_NONE:
+			/*
+			 * If neither hibernation nor partial power down are supported,
+			 * clock gating is used to save power.
+			 */
+			if (!hsotg->params.no_clock_gating)
+				dwc2_gadget_enter_clock_gating(hsotg);
+		}
+	}
+	spin_unlock_irqrestore(&hsotg->lock, flags);
+
 	return 0;
 }
 
--- a/drivers/usb/dwc2/gadget.c
+++ b/drivers/usb/dwc2/gadget.c
@@ -3254,6 +3254,45 @@ static void dwc2_hsotg_epint(struct dwc2
 	}
 }
 
+static void dwc2_gadget_exit_lp(struct dwc2_hsotg *hsotg)
+{
+	if (hsotg->lx_state != DWC2_L2)
+		return;
+
+	switch (hsotg->params.power_down) {
+	case DWC2_POWER_DOWN_PARAM_PARTIAL:
+		if (dwc2_exit_partial_power_down(hsotg, 0, true))
+			dev_err(hsotg->dev, "exit partial_power_down failed\n");
+		return;
+
+	case DWC2_POWER_DOWN_PARAM_NONE:
+		if (!hsotg->params.no_clock_gating && hsotg->bus_suspended)
+			dwc2_gadget_exit_clock_gating(hsotg, 0);
+	}
+}
+
+static void dwc2_gadget_enter_lp(struct dwc2_hsotg *hsotg)
+{
+	/* Only enter LP when in L0 state, or when the core has been disconnected / stopped */
+	if (hsotg->lx_state != DWC2_L0 && hsotg->lx_state != DWC2_L3)
+		return;
+
+	switch (hsotg->params.power_down) {
+	case DWC2_POWER_DOWN_PARAM_PARTIAL:
+		if (dwc2_enter_partial_power_down(hsotg))
+			dev_err(hsotg->dev, "enter partial_power_down failed\n");
+		return;
+
+	case DWC2_POWER_DOWN_PARAM_NONE:
+		/*
+		 * If neither hibernation nor partial power down are supported,
+		 * clock gating is used to save power.
+		 */
+		if (!hsotg->params.no_clock_gating)
+			dwc2_gadget_enter_clock_gating(hsotg);
+	}
+}
+
 /**
  * dwc2_hsotg_irq_enumdone - Handle EnumDone interrupt (enumeration done)
  * @hsotg: The device state.
@@ -3678,21 +3717,6 @@ void dwc2_hsotg_core_init_disconnected(s
 
 void dwc2_hsotg_core_disconnect(struct dwc2_hsotg *hsotg)
 {
-	/*
-	 * If controller is in partial power down state or in clock gating mode, it must
-	 * exit from that state before being disconnected, and detached.
-	 */
-	if (hsotg->lx_state == DWC2_L2) {
-		/* No need to check the return value as registers are not being restored. */
-		if (hsotg->in_ppd)
-			dwc2_exit_partial_power_down(hsotg, 0, false);
-
-		if (hsotg->params.power_down ==
-		    DWC2_POWER_DOWN_PARAM_NONE && hsotg->bus_suspended &&
-		    !hsotg->params.no_clock_gating)
-			dwc2_gadget_exit_clock_gating(hsotg, 0);
-	}
-
 	/* set the soft-disconnect bit */
 	dwc2_set_bit(hsotg, DCTL, DCTL_SFTDISCON);
 }
@@ -4417,6 +4441,9 @@ static int dwc2_hsotg_ep_disable_lock(st
 	unsigned long flags;
 	int ret;
 
+	/* Need to exit power saving to disable endpoints. */
+	dwc2_gadget_exit_lp(hsotg);
+
 	spin_lock_irqsave(&hsotg->lock, flags);
 	ret = dwc2_hsotg_ep_disable(ep);
 	spin_unlock_irqrestore(&hsotg->lock, flags);
@@ -4685,6 +4712,9 @@ static int dwc2_hsotg_udc_start(struct u
 			goto err;
 	}
 
+	/* Exit low power mode, before starting */
+	dwc2_gadget_exit_lp(hsotg);
+
 	if (hsotg->params.reset_phy_on_start) {
 		/*
 		 * In case the gadget has been stopped earlier, and
@@ -4737,11 +4767,7 @@ static int dwc2_hsotg_udc_stop(struct us
 
 	device_set_wakeup_capable(&gadget->dev, false);
 
-	/* Exit clock gating when driver is stopped. */
-	if (hsotg->params.power_down == DWC2_POWER_DOWN_PARAM_NONE &&
-	    hsotg->bus_suspended && !hsotg->params.no_clock_gating) {
-		dwc2_gadget_exit_clock_gating(hsotg, 0);
-	}
+	dwc2_gadget_exit_lp(hsotg);
 
 	/* all endpoints should be shutdown */
 	for (ep = 1; ep < hsotg->num_of_eps; ep++) {
@@ -4763,6 +4789,8 @@ static int dwc2_hsotg_udc_stop(struct us
 	if (!IS_ERR_OR_NULL(hsotg->uphy))
 		otg_set_peripheral(hsotg->uphy->otg, NULL);
 
+	dwc2_gadget_enter_lp(hsotg);
+
 	if (hsotg->dr_mode == USB_DR_MODE_PERIPHERAL)
 		dwc2_lowlevel_hw_disable(hsotg);
 
@@ -4879,6 +4907,11 @@ static int dwc2_hsotg_pullup(struct usb_
 		return 0;
 	}
 
+	/* Need to exit low power, before (dis-)connecting */
+	spin_lock_irqsave(&hsotg->lock, flags);
+	dwc2_gadget_exit_lp(hsotg);
+	spin_unlock_irqrestore(&hsotg->lock, flags);
+
 	if (hsotg->params.stm32_has_batt_chg_det && is_on) {
 		ret = stm32mp2_usb2phy_batt_chg_det(hsotg);
 		if (ret)
@@ -4916,12 +4949,7 @@ static int dwc2_hsotg_vbus_session(struc
 	 * If controller is in partial power down state, it must exit from
 	 * that state before being initialized / de-initialized
 	 */
-	if (hsotg->lx_state == DWC2_L2 && hsotg->in_ppd)
-		/*
-		 * No need to check the return value as
-		 * registers are not being restored.
-		 */
-		dwc2_exit_partial_power_down(hsotg, 0, false);
+	dwc2_gadget_exit_lp(hsotg);
 
 	if (is_active) {
 		hsotg->op_state = OTG_STATE_B_PERIPHERAL;
@@ -5198,6 +5226,8 @@ int dwc2_gadget_init(struct dwc2_hsotg *
 	int epnum;
 	int ret;
 
+	pm_runtime_get(hsotg->dev);
+
 	/* Dump fifo information */
 	dev_dbg(dev, "NonPeriodic TXFIFO size: %d\n",
 		hsotg->params.g_np_tx_fifo_size);
@@ -5804,6 +5834,8 @@ int dwc2_gadget_enter_partial_power_down
 	hsotg->lx_state = DWC2_L2;
 	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
 
+	pm_runtime_put(hsotg->dev);
+	hsotg->rpm_suspended = true;
 	dev_dbg(hsotg->dev, "Entering device partial power down completed.\n");
 
 	return ret;
@@ -5833,6 +5865,16 @@ int dwc2_gadget_exit_partial_power_down(
 
 	dev_dbg(hsotg->dev, "Exiting device partial Power Down started.\n");
 
+	if (hsotg->rpm_suspended) {
+		/*
+		 * Rely on rpm_suspended flag, to resume once, as gadget/host
+		 * exit_partial_power_down routines may get called several times,
+		 * with rem_wakeup.
+		 */
+		pm_runtime_get(hsotg->dev);
+		hsotg->rpm_suspended = false;
+	}
+
 	pcgcctl = dwc2_readl(hsotg, PCGCTL);
 	pcgcctl &= ~PCGCTL_STOPPCLK;
 	dwc2_writel(hsotg, pcgcctl, PCGCTL);
@@ -5921,6 +5963,9 @@ void dwc2_gadget_enter_clock_gating(stru
 	hsotg->lx_state = DWC2_L2;
 	hsotg->bus_suspended = true;
 	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
+
+	pm_runtime_put(hsotg->dev);
+	hsotg->rpm_suspended = true;
 }
 
 /*
@@ -5938,6 +5983,15 @@ void dwc2_gadget_exit_clock_gating(struc
 
 	dev_dbg(hsotg->dev, "Exiting device clock gating.\n");
 
+	if (hsotg->rpm_suspended) {
+		/*
+		 * Rely on rpm_suspended flag, to resume once, as gadget/host exit_clock_gating
+		 * routines may get called several times, with rem_wakeup.
+		 */
+		pm_runtime_get(hsotg->dev);
+		hsotg->rpm_suspended = false;
+	}
+
 	/* Clear the Gate hclk. */
 	pcgctl = dwc2_readl(hsotg, PCGCTL);
 	pcgctl &= ~PCGCTL_GATEHCLK;
--- a/drivers/usb/dwc2/hcd.c
+++ b/drivers/usb/dwc2/hcd.c
@@ -5815,6 +5815,8 @@ int dwc2_host_enter_partial_power_down(s
 	hsotg->lx_state = DWC2_L2;
 	hsotg->bus_suspended = true;
 
+	pm_runtime_put(hsotg->dev);
+	hsotg->rpm_suspended = true;
 	dev_dbg(hsotg->dev, "Entering host partial power down completed.\n");
 
 	return ret;
@@ -5841,6 +5843,15 @@ int dwc2_host_exit_partial_power_down(st
 
 	dev_dbg(hsotg->dev, "Exiting host partial power down started.\n");
 
+	/*
+	 * Rely on rpm_suspended, to resume once, as gadget/host partial_power_down
+	 * routines may get called several times, with rem_wakeup.
+	 */
+	if (hsotg->rpm_suspended) {
+		pm_runtime_get(hsotg->dev);
+		hsotg->rpm_suspended = false;
+	}
+
 	pcgcctl = dwc2_readl(hsotg, PCGCTL);
 	pcgcctl &= ~PCGCTL_STOPPCLK;
 	dwc2_writel(hsotg, pcgcctl, PCGCTL);
@@ -5940,6 +5951,9 @@ void dwc2_host_enter_clock_gating(struct
 
 	hsotg->bus_suspended = true;
 	hsotg->lx_state = DWC2_L2;
+
+	pm_runtime_put(hsotg->dev);
+	hsotg->rpm_suspended = true;
 }
 
 /**
@@ -5957,6 +5971,15 @@ void dwc2_host_exit_clock_gating(struct
 
 	dev_dbg(hsotg->dev, "Exiting host clock gating.\n");
 
+	if (hsotg->rpm_suspended) {
+		/*
+		 * Rely on rpm_suspended, to resume once, as gadget/host exit_clock_gating
+		 * routines may get called several times, with rem_wakeup.
+		 */
+		pm_runtime_get(hsotg->dev);
+		hsotg->rpm_suspended = false;
+	}
+
 	/* Clear the Gate hclk. */
 	pcgctl = dwc2_readl(hsotg, PCGCTL);
 	pcgctl &= ~PCGCTL_GATEHCLK;
--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -663,6 +663,14 @@ static int dwc2_driver_probe(struct plat
 	if (retval)
 		return retval;
 
+	retval = pm_runtime_set_active(&dev->dev);
+	if (retval)
+		return dev_err_probe(&dev->dev, retval, "Failed to activate pm runtime\n");
+
+	retval = devm_pm_runtime_enable(&dev->dev);
+	if (retval)
+		return dev_err_probe(&dev->dev, retval, "Failed to enable pm runtime\n");
+
 	spin_lock_init(&hsotg->lock);
 
 	hsotg->vbus_supply = devm_regulator_get_optional(hsotg->dev, "vbus");
@@ -1010,6 +1018,10 @@ static int __maybe_unused dwc2_resume(st
 	}
 	dwc2->phy_off_for_suspend = false;
 
+	pm_runtime_disable(dev);
+	pm_runtime_set_active(dev);
+	pm_runtime_enable(dev);
+
 	if (dwc2->params.activate_stm_id_vb_detection) {
 		unsigned long flags;
 		u32 ggpio, gotgctl;
